docs: add daemon docs (#837)
VeryGoodOpenSource/dart_frog: docs/docs/advanced/daemon.md

import TOCInline from '@theme/TOCInline';

# 🧰 Daemon

> Dart frog daemon is a standing process that, during its lifetime, will be used by first and

Reviewer: ```suggestion
Dart Frog daemon is a standing process that, during its lifetime, will be used by first and
```

Reviewer: sorry last one promise!!

> Dart frog daemon is a standing process that, during its lifetime, will be used by first and
> third-party tools to manage, build, and diagnose dart frog projects.

Reviewer: ```suggestion
Dart frog daemon is a standing process that, during its lifetime, will be used by first and
third-party tools to manage, build, and diagnose Dart Frog projects.
```

By design, the daemon is able to manage multiple projects simultaneously; it can also run multiple
application instances of the same project if necessary.

To start using it, install the Dart Frog CLI and run the `dart_frog daemon` command. Once running, communicating with it can be done via [JSON-RPC](https://www.jsonrpc.org/) over stdin/stdout to receive and send messages.

> The messages can be sorted in three types:

Reviewer 2: ```suggestion
There are three types of messages:
```

import TOCInline from '@theme/TOCInline';

# 🧰 Daemon

Dart frog daemon is a standing process that, during its lifetime, will be used by first and
third-party tools to manage, build, and diagnose dart frog projects.

> By design, a daemon is able to manage multiple projects simultaneously. The daemon can run multiple
> application instances of the same project if necessary.
> 
> The daemon can be started by running the `dart_frog daemon` command. The daemon
> uses [JSON-RPC](https://www.jsonrpc.org/) over stdin/stdout to receive and send messages.

Reviewer 3: **Nit**: I feel that starting three consecutive sentences with "The deamon" might make readers feel bombarded with daemon facts. Maybe we can use other connectors. Here is a possible suggestion:

```suggestion
By design, the daemon is able to manage multiple projects simultaneously. It can also run multiple
application instances of the same project if necessary.

To start using it,  install the Dart Frog CLI and run the `dart_frog daemon` command. Once running, communicating with it can be done via [JSON-RPC](https://www.jsonrpc.org/) over stdin/stdout to receive and send messages.
```

The messages can be sorted in three types:

- **Request**: A request is a message sent by a client to the daemon. The daemon will process the
  request and send a response back to the client. A request is essentially a method invocation.
- **Response**: A response is a message sent by the daemon to a client in response to a request.
- **Event**: An event is a message sent by the daemon to a client. The daemon will send an event to
  a client when something happens, for example, when a running dev server stops.

Every request should be met with a response as soon as possible so the caller can work with
timeouts. The daemon will send events to the client as they happen.

---

#### Usage example

```json
$ dart_frog daemon

// ready event sent via stdout
[{"event":"daemon.ready","params":{"version":"0.0.1","processId":75941}}]

// request inserted via stdin
[{"method": "daemon.requestVersion", "id": "12"}]

// response sent via stdout
[{"id":"12","result":{"version":"0.0.1"}}]

```

> The `id` field on the request is used to match the request with the response. As the client sets it arbitrarily, the client is responsible for ensuring that the id is unique.

Reviewer 3: ```suggestion
The `id` field on the request is used to match the request with the response. As the client sets it arbitrarily, the client is responsible for ensuring that all request ids are unique.
```

---

> To organize the accepted requests and its parameters as well as events, there are the "domains". A
> domain is a group of related requests and events.
> 
> The domains are:

Reviewer 3: ```suggestion
Internally, the daemon uses "domains" to logically group related requests and events. These are:
```
